New module proxmox_kvm

helldorado opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- New Module Pull Request
##### COMPONENT NAME

<!--- proxmox_kvm -->

proxmox_kvm 
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible --version                                                                                                                                                                                                                                                                                            
ansible 2.1.2.0
  config file =
  configured module search path = Default w/o overrides
```
##### SUMMARY

<!--- Describe the change, including rationale and design decisions -->

Management of Qemu(KVM) Virtual Machines in Proxmox
Allows you to create/delete/stop Qemu(KVM) Virtual Machines in Proxmox VE cluster.

<!---
If you are fixing an existing issue, please include "Fixes #nnnn" in your commit
message and your description; but you should still explain what the change does.
-->

<!--- Paste verbatim command output below, e.g. before and after your change -->

```
ansible-playbook test.yml                                                                                                                                                                                                                                                                      PLAY [localhost] ***************************************************************

TASK [proxmox-build : CREATE VM] ***************************************************
changed: [localhost]

TASK [proxmox-build : debug] *******************************************************
ok: [localhost] => {
    "qemu": {
        "changed": true,
        "devices": {
            "virtio0": "VMs:130/vm-130-disk-1.raw"
        },
        "mac": {
            "net0": "26:46:D5:C2:4B:AC"
        },
        "msg": "VM helldorado with vmid 130 deployed",
        "vmid": 130
    }
}

TASK [proxmox-build : START VM] ****************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0
```

gregdek commented:
Thanks @helldorado for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

MartinVerges commented:
shipit

morph027 commented:
Exactly was i was looking for...unfortunately i can't create new a vm (starting, stopping works fine)

Latest Proxmox VE 4.3

```
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "acpi": true, 
            "agent": null, 
            "api_host": "proxmox", 
            "api_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "api_user": "root@pam", 
            "args": null, 
            "autostart": false, 
            "balloon": 0, 
            "bios": null, 
            "boot": "cnd", 
            "bootdisk": null, 
            "cores": 1, 
            "cpu": "kvm64", 
            "cpulimit": null, 
            "cpuunits": 1000, 
            "delete": null, 
            "description": null, 
            "digest": null, 
            "force": false, 
            "freeze": null, 
            "hostpci": null, 
            "hotplug": null, 
            "hugepages": null, 
            "ide": null, 
            "keyboard": null, 
            "kvm": true, 
            "localtime": null, 
            "lock": null, 
            "machine": null, 
            "memory": 512, 
            "migrate_downtime": null, 
            "migrate_speed": null, 
            "name": "testmehard", 
            "net": null, 
            "node": "proxmox", 
            "numa": null, 
            "onboot": true, 
            "ostype": "l26", 
            "parallel": null, 
            "protection": null, 
            "reboot": null, 
            "revert": null, 
            "sata": null, 
            "scsi": null, 
            "scsihw": null, 
            "serial": null, 
            "shares": null, 
            "skiplock": null, 
            "smbios": null, 
            "sockets": 1, 
            "startdate": null, 
            "startup": null, 
            "state": "present", 
            "tablet": false, 
            "tdf": null, 
            "template": false, 
            "timeout": 30, 
            "validate_certs": false, 
            "vcpus": null, 
            "vga": "std", 
            "virtio": null, 
            "vmid": null, 
            "watchdog": null
        }, 
        "module_name": "proxmox_kvm"
    }, 
    "msg": "creation of qemu VM testmehard with vmid 111 failed with exception: 400 Bad Request"
}
```

 Unfortunately debugging on PVE is quite limited. I can't get any info why PVE is complaining.

Doing the same with plain proxmoxer works fine.

morph027 commented:
Ok, looks like i've boiled down the problem:

Commenting all params in `create_vm()` ([line 893-942](https://github.com/helldorado/ansible-modules-extras/blob/b19c3390461b281883349e75940ecdece94d1947/cloud/misc/proxmox_kvm.py#L893-L942)) and uncommenting step by step yields the error at the `force` flag.

Looking at the [API documentation](http://pve.proxmox.com/pve-docs/api-viewer/index.html) (`nodes - {node} - qemu - post`), this should be allowed to overwrite vms:

> | Name | Type | Format | Description |
> | --- | --- | --- | --- |
> | force | boolean | boolean | Allow to overwrite existing VM. |

This happens when creating a new vm with a name which has not been taken and no vmid set (get from cluster).

morph027 commented:
Also found a typo in `vga` param which leads to refuse vm creation: https://github.com/helldorado/ansible-modules-extras/pull/1

morph027 commented:
Regarding the `force` issue:

``` bash
# qm create $(pvesh get /cluster/nextid | grep -oE '[0-9]*') --name something --force
200 OK
400 Parameter verification failed.
archive: missing property - 'force' requires this property
qm create <vmid> [OPTIONS]
```

My best guess from what i see and experienced with PVE so far, the force flag only applies to restore from archive (i can't remind that it's even possible to create a new vm from scratch and replace an existing one).

[Asking for help/clarification in PVE forums](https://forum.proxmox.com/threads/api-force-flag.30224/)...

helldorado commented:
@morph027  this issue as solved. Let me know when you retry it.

``` bash
root@zbxhyp1:~ # pveversion
pve-manager/4.3-1/e7cdc165 (running kernel: 4.2.8-1-pve)
```

``` yaml
ansible-playbook test.yml

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [proxmox-build : CREATE VM] ***************************************************
changed: [localhost]

TASK [proxmox-build : debug] *******************************************************
ok: [localhost] => {
    "qemu": {
        "changed": true,
        "devices": {
            "virtio0": "VMs_zbxhyp1:110/vm-110-disk-1.raw",
            "virtio1": "VMs_zbxhyp1:110/vm-110-disk-2.qcow2"
        },
        "mac": {
            "net0": "46:B4:15:B9:96:DD"
        },
        "msg": "VM kropta with vmid 110 deployed",
        "vmid": 110
    }
}

TASK [proxmox-build : Create host in foreman] *************************************************
ok: [localhost]

TASK [proxmox-build : START VM on vlan61 and boot on PXE] ****************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=5    changed=2    unreachable=0    failed=0
```

morph027 commented:
@helldorado : Confirmed, works ;) Thank you.

gundalow commented:
Hi,

Looks like there is a small tweak needed to support Python 3

```
2016-11-10 09:57:45 *** Error compiling './cloud/misc/proxmox_kvm.py'...
2016-11-10 09:57:45   File "./cloud/misc/proxmox_kvm.py", line 644
2016-11-10 09:57:45     except Exception, e:
2016-11-10 09:57:45                     ^
2016-11-10 09:57:45 SyntaxError: invalid syntax
```

The following guide should explain what's needed http://docs.ansible.com/ansible/developing_modules_python3.html

Thanks

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+options:
+  acpi:
+    description:
+      - enable/disable ACPI
```
All `description:` need to be full sentences, e.g. capital letter and full stop.

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+    type: boolean
+  agent:
+    description:
+      - enable/disable Qemu GuestAgent
```
Since you are referring to a boolean I wonder if the wording for this (and others) should be along the lines of

> Specify if the Qemu GuestAgent should be enabled.

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+  args:
+    description:
+      - pass arbitrary arguments to kvm
+    notes: this option is for experts only
```
I don't believe `notes:` are supported under `option` arguments, so this should be added onto the description line.

gregdek commented:
Thanks @helldorado for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+    required: false
+  autostart:
+    description:
+      - automatic restart after crash (currently ignored)
```
Why is this ignored?
Should this option be added in the future when it's fully working?

morph027 replied:
Looks like upstream project (Proxmox) does not currently use this, the string itself is copied directly from their API description at http://pve.proxmox.com/pve-docs/api-viewer/index.html

gundalow replied:
Should it be removed from the Ansible docs then?

morph027 replied:
I've prepared a PR for @helldorado with the following:

```
Specify, if the VM should be automatically restarted after crash (currently ignored in PVE API).
```

helldorado replied:
@gundalow thank's for all your feedback, @morph027 PR have done some of your preconisation.

This option is ignored by proxmox's API. I suppose it will be working in a future release of proxmox or of it's API, and this won't need any update on this module.

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+
+# import module snippets
+from ansible.module_utils.basic import *
+main()
```
Needs rewriting as:

```
if __name__ == '__main__':
    main()
```

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+  boot:
+    description:
+      - boot on floppy (a), hard disk (c), CD-ROM (d), or network (n)
+    notes: You can combine to set order.
```
Again, `notes` not supported here

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+    type: A hash/dictionary defining ide
+  keyboard:
+    description:
+      - sets the keybord layout for vnc server
```
keybord > keyboard
vnc -> VNC

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+    default: present
+  tablet:
+    description:
+      - enable/disable the usb tablet device
```
USB

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+        "virtio2": "VMs:115/vm-115-disk-2.raw"
+      }'
+mac:
+    description: List of mac address created and net[n] rattached. Userfull when you want to use Provisionning system like Foreman via PXE.
```
rattached -> attached
Userfull -> useful
Provisionning -> provision

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+        try:
+          vm = proxmox.nodes(node).qemu(vmid).config.get()
+        except Exception, e:
+          module.fail_json(msg='Getting informations for VM with vmid = %s failed with exception: %s' % (vmid, e))
```
informations -> information

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1020 @@
+  # If vmid not set get the Next VM id from ProxmoxAPI
+  # If vm name is set get the VM id from ProxmoxAPI
+  if module.params['vmid'] is not None:
+   vmid = module.params['vmid']
```
Indentation in this section doesn't look right

helldorado replied:
Thank's again @gundalow

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1021 @@
+  api_password:
+    description:
+      - Specify the password to authenticate with.
+      - you can use PROXMOX_PASSWORD environment variable
```
You can use C(PROXMOX_PASSWORD) environment variable.

helldorado commented:
From the latest commit  [476c1c6](https://github.com/ansible/ansible-modules-extras/pull/3292/commits/476c1c62272009368fe021080a7d4baec7c43f8e),  `current`  **state** was added.

``` yaml
ansible-playbook test.yml
PLAY [localhost] ***************************************************************

TASK [proxmox-build : VM STATUS] ***************************************************
ok: [localhost]

TASK [proxmox-build : debug] *******************************************************
ok: [localhost] => {
    "qemu": {
        "changed": false,
        "msg": "VM kropta with vmid = 110 is running",
        "status": "running"
    }
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```

helldorado commented:
ready_for_review

gregdek commented:
Thanks @helldorado for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

morph027 commented:
Did you try to create VMs after your introduction of `state: current` ?

If i do try now, i get the following:

```
TASK [Create KVM] ****************************************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "creation of qemu VM foobar with vmid 111 failed with exception: local variable 'results' referenced before assignment"}
```

The VM however gets created successfully, but the playbook bails out immediately.

MihaiMoldovanu commented:
Hi , I created this patch that allows cloning kvm templates ( or full machines ) using a playbook like this:

---

# clone VMs on proxmox
- hosts: localhost
  connection: local
  tasks:
  - proxmox_kvm:
    # connect params to proxmox
    api_user    : tfm@pve
    api_password: XXXX
    api_host    : 111.111.111.11
    # qemu template id.
    clone       : 555
    # new vm id.
    vmid        : 666
    node        : proxmox2
    name        : onetest
    storage     : 'stor2'
    format      : 'qcow2'
    timeout     : 300

# 

diff -Naur /usr/lib/python2.7/dist-packages/ansible/modules/core/cloud/misc/proxmox_kvm.py proxmox_kvm.py 
--- /usr/lib/python2.7/dist-packages/ansible/modules/core/cloud/misc/proxmox_kvm.py 2016-11-15 01:51:21.736857612 +0200
+++ proxmox_kvm.py  2016-11-10 20:31:03.504478194 +0200
@@ -644,10 +644,7 @@
   if module.params['api_user'] != "root@pam" and module.params['skiplock'] is not None:
     module.fail_json(msg='skiplock parameter require root@pam user. ')
-  if module.params['clone'] == 'None':
-    taskid = getattr(proxmox_node, VZ_TYPE).create(vmid=vmid, name=name, memory=memory, cpu=cpu, cores=cores, sockets=sockets, **kwargs)
-  else:
-    taskid = proxmox_node.qemu(module.params['clone']).clone.post(newid=vmid, full='1', name=name, storage=module.params['storage'], format=module.params['format'])
-  taskid = getattr(proxmox_node, VZ_TYPE).create(vmid=vmid, name=name, memory=memory, cpu=cpu, cores=cores, sockets=sockets, **kwargs)
  
  while timeout:
     if ( proxmox_node.tasks(taskid).status.get()['status'] == 'stopped'
  @@ -757,9 +754,6 @@
     vga = dict(default='std', choices=['std', 'cirrus', 'vmware', 'qxl', 'serial0', 'serial1', 'serial2', 'serial3', 'qxl2', 'qxl3', 'qxl4']),
     virtio = dict(type='dict', default=None),
     vmid = dict(type='int', default=None),
-      clone = dict(type='int', default=None),
-      storage = dict(type='str'),
-      format = dict(type='str'),
     watchdog = dict(),
   )
  )
  ===============================================================================
  root@hq:/1.Work/ansible-proxmox# ansible-playbook deploy_vms.yaml -vv
  Using /etc/ansible/ansible.cfg as config file
  1 plays in deploy_vms.yaml

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [proxmox_kvm] *************************************************************
changed: [localhost] => {"changed": true, "devices": {}, "mac": {"net0": "12:CA:F6:65:C0:B1"}, "msg": "VM core07 with vmid 556 deployed", "vmid": 556}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0   

# 

I hope you find it useful . 

Regards, 
Mihai

helldorado commented:
@MihaiMoldovanu greats, Can you post a pull request on my branch? The clone can be quite interesting.
Let us wait, however, for a second shipit before merging here.

MihaiMoldovanu commented:
@helldorado  Done. Pull requested to your branch. I added documentation inside too.

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1055 @@
+  balloon:
+    description:
+      - Specify the amount of RAM for the VM in MB.
+      - Using zero disables the ballon driver.
```
ballon -> balloon

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1055 @@
+    type: string
+  bootdisk:
+    description:
+      - Specify if booting from specified disk is enabled.
```
"Boot from the specified disk."

Can you list multiple disks for it to try in order? May be worth clarifying that point.

morph027 replied:
Nope, just one disk. Again, this is doc string provided by the proxmox api itself.

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1055 @@
+    type: integer
+  validate_certs:
+    description:
+      - Enables/disables HTTPS certificate verification for API access.
```
Can you please change this to the following so it's consistent with other modules and the security implication is clearer:
"If C(no), SSL certificates will not be validated. This should only be used on personally controlled sites using self-signed certificates."

gundalow commented on cloud/misc/proxmox_kvm.py:
```diff
@@ -0,0 +1,1055 @@
+
+  taskid = getattr(proxmox_node, VZ_TYPE).create(vmid=vmid, name=name, memory=memory, cpu=cpu, cores=cores, sockets=sockets, **kwargs)
+
+  while timeout:
```
Indentation in this section looks wrong

gundalow commented:
@helldorado Thanks for the updates
I'll just wait for approval from @MihaiMoldovanu before merging this

MihaiMoldovanu commented:
@gundalow : no more modification from me at this time . Thanks

gundalow commented:
Squashed and merged. Thanks for all the reviews.

rompic commented:
+1. we were looking for this^^

gundalow commented:
@rompic Ace, You either run from source, or better yet create a `library` directory (at the same level as `roles`, `group_vars` etc) and put https://raw.githubusercontent.com/ansible/ansible-modules-extras/devel/cloud/misc/proxmox_kvm.py in there

helldorado commented:
@rompic here is an sample tree. 

```bash
.
├── README.md
├── defaults
│   └── main.yml
├── examples
│   ├── hosts
│   └── playbook.yml
├── files
├── handlers
│   └── main.yml
├── library
│   ├── foreman_host.py
│   ├── irc.py
│   └── proxmox_kvm.py
├── meta
│   └── main.yml
├── tasks
│   ├── container.yml
│   ├── main.yml
│   ├── physical.yml
│   ├── proxmox.yml
│   └── virtual.yml
├── templates
├── tests
│   ├── integration
│   ├── inventory
│   └── test.yml
└── vars
    └── main.yml
```

rompic commented:
We successfully tested from source today. Also looked into the clone pull request. Lgtm! Nice work

seljuck commented:
Hello,
I just attempted to use this module without success. I copied it to my /library directory. 

I'm using Ansible 2.2.0.0 on Ubuntu Mate 16.10

Here's the playbook
```
- hosts: proxmox
  vars:

  tasks:
  - name: test proxmox connection
    proxmox_kvm:
      api_user    : root@pam
      api_password: password
      api_host    : proxmox1
      name        : example
      node        : proxmox1
      vmid        : 102
      state       : current
```

Here's the error message I'm getting.

```
TASK [test proxmox connection] *************************************************
task path: /home/administrator/example/test-proxmox.yaml:8
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 119, in run
    res = self._execute()
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/task_executor.py", line 490, in _execute
    result = self._handler.run(task_vars=variables)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/action/normal.py", line 33, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars))
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/action/__init__.py", line 565, in _execute_module
    (module_style, shebang, module_data, module_path) = self._configure_module(module_name=module_name, module_args=module_args, task_vars=task_vars)
  File "/usr/local/lib/python2.7/dist-packages/ansible/plugins/action/__init__.py", line 153, in _configure_module
    task_vars=task_vars, module_compression=self._play_context.module_compression)
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/module_common.py", line 772, in modify_module
    (module_data, module_style, shebang) = _find_snippet_imports(module_name, module_data, module_path, module_args, task_vars, module_compression)
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/module_common.py", line 640, in _find_snippet_imports
    recursive_finder(module_name, module_data, py_module_names, py_module_cache, zf)
  File "/usr/local/lib/python2.7/dist-packages/ansible/executor/module_common.py", line 441, in recursive_finder
    tree = ast.parse(data)
  File "/usr/lib/python2.7/ast.py", line 37, in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
TypeError: compile() expected string without null bytes

fatal: [proxmox1.example.com]: FAILED! => {
    "failed": true, 
    "msg": "Unexpected failure during module execution.", 
    "stdout": ""
}
        to retry, use: --limit @/home/administrator/example/test-proxmox.retry

```

morph027 commented:
Do you have the ```proxmoxer``` module installed at your proxmox host?

> Requirements (on host that executes module)
>        proxmoxer
>        requests

You can also run it locally, there's no need to run it on the proxmox host, as it does API calls:

```sudo pip install proxmoxer requests```

```
- hosts: localhost
  gather_facts: no
  vars:

  tasks:
  - name: test proxmox connection
    proxmox_kvm:
      api_user    : root@pam
      api_password: password
      api_host    : proxmox1
      name        : example
      node        : proxmox1
      vmid        : 102
      state       : current
```

seljuck commented:
Yes I do have proxmoxer and requests installed on the Ansible Controller PC, NOT on the Proxmox host. 

Thanks for the help, 
Daniel 





Do you have the proxmoxer module installed at your proxmox host? 

You can also run it locally, there's no need to run it on the proxmox host, as it does API calls: 

sudo pip install proxmoxer

helldorado commented:
@seljuck 

Please move your issue here => https://github.com/ansible/ansible/issues/

The ansible-extras has moved to ansible. For more info =>  https://docs.ansible.com/ansible/dev_guide/repomerge.html

seljuck commented:
This issue is now moved to https://github.com/ansible/ansible/issues/20316

robinro commented:
@helldorado How is this different from the `proxmox` module?
It would be nice to have some explanation in the docs of features that only this module or only `proxmox` can do.

morph027 commented:
Looks like the proxmox module is for containers only...
